clang/lib/CodeGen/CodeGenFunction.cpp

```diff
@@ -694,7 +694,7 @@ void CodeGenFunction::EmitOpenCLKernelMetadata(const FunctionDecl *FD,
 
   if (const SYCLIntelNoGlobalWorkOffsetAttr *A =
           FD->getAttr<SYCLIntelNoGlobalWorkOffsetAttr>()) {
-    if (A->getEnabled())
+    if (A->getValue())
```

Comment:
FWIW, this is checking that the attribute has a non-null `Expr*`, not what the value of the expression is. Should you be checking for non-zero vs zero here?

Reply:
According to the documentation *https://github.com/smanna12/llvm/blob/83e0f1332cbfcbc73058b11b9c8fb7c260168652/clang/include/clang/Basic/AttrDocs.td#L2334),

If 1, compiler doesn't use the global work offset values for
the device function. Valid values are 0 and 1. If used without argument, value
of 1 is set implicitly. 

I have checked the attribute has non-null EXpr * instead of value of the expression to keep the IR same (clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp) as we had before the template support (PR https://github.com/intel/llvm/pull/1090)

Reply:
So it sounds like this isn't correct currently (and that suggests there's test coverage that's missing or incorrect) -- `A->getValue()` will always be `true` even if the user wrote `[[intel::no_global_work_offset(0)]]` because it's testing the `Expr*` is not null (and it's always nonnull) rather than the expression's value.

I think the code here should be:
```
const Expr *Arg = A->getValue();
assert(Arg && "Got an unexpected null argument");
llvm::Optional<llvm::APSInt> OptVal = Arg->getIntegerConstantExpr(getContext());
assert(OptVal && "Got something other than an integer constant expression");
if (*OptVal)
  Fn->setMetadata(...);
```

Reply:
@AaronBallman, thanks for the suggestion. I have updated codegen part.

One issue is here that we can not use the check if (*OptVal) here. This will cause an error.
 error: could not convert ‘OptVal.llvm::Optional<llvm::APSInt>::operator*()’ from ‘llvm::APSInt’ to ‘bool’
     if (*OptVal)
         ^~~~~~~

Reply:
Oh, that's a think-o on my part, thanks for correcting it. :-)